Backpack/src/main/java/net/skyscanner/backpack/calendar/presenter/MonthAdapter.kt

```diff
@@ -17,81 +17,134 @@
 
 import android.content.Context
+import android.database.DataSetObserver
 import android.view.View
 import android.view.ViewGroup
-import android.widget.AbsListView.LayoutParams
 import android.widget.BaseAdapter
 import net.skyscanner.backpack.calendar.model.CalendarDrawingParams
 import net.skyscanner.backpack.calendar.view.MonthView
 import org.threeten.bp.LocalDate
+import org.threeten.bp.Period
 
 internal class MonthAdapter(
   private val context: Context,
   private val controller: BpkCalendarController
 ) : BaseAdapter(), MonthView.OnDayClickListener {
 
-    private var selectedDay: LocalDate? = null
-        set(value) {
-            field = value
-            notifyDataSetChanged()
-        }
-
-    override fun onDayClick(view: MonthView?, day: LocalDate) {
-        onDayTapped(day)
+  private var selectedDay: LocalDate? = null
+    set(value) {
+      field = value
+      notifyDataSetChanged()
     }
 
-  override fun getCount() =
-    (controller.endDate.year - controller.startDate.year) * MONTHS_IN_YEAR +
-      controller.endDate.month.value - controller.startDate.month.value + 1
+  private var positionMetadata = computeMetadata()
 
-    override fun getItem(position: Int) = null
+  init {
+    registerDataSetObserver(object : DataSetObserver() {
+      override fun onChanged() {
+        positionMetadata = computeMetadata()
+      }
+    })
+  }
 
-    override fun getItemId(position: Int) = position.toLong()
+  override fun onDayClick(view: MonthView?, day: LocalDate) {
+    onDayTapped(day)
+  }
 
-    override fun hasStableIds() = true
+  override fun getCount() = positionMetadata.size
 
-    @Suppress("UNCHECKED_CAST")
-    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
-        val view: MonthView
+  override fun getItem(position: Int) = null
 
-        if (convertView != null) {
-            view = convertView as MonthView
-        } else {
-            view = createMonthView(context).apply {
-                layoutParams = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT)
-                isClickable = true
-            }
-            view.onDayClickListener = this@MonthAdapter
-        }
+  override fun getItemId(position: Int) = position.toLong()
 
-        val positionWithStart = position + controller.startDate.month.value - 1
-        val month = positionWithStart % MONTHS_IN_YEAR + 1
-        val year = positionWithStart / MONTHS_IN_YEAR + controller.startDate.year
-        val selectedDay = selectedDay?.let {
-          if (isSelectedDayInMonth(it, year, month)) it.dayOfMonth else null
-        }
+  override fun hasStableIds() = true
 
-        view.reuse()
+  @Suppress("UNCHECKED_CAST")
+  override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
+    val metadata = positionMetadata[position]
+    val month = metadata.month
+    val year = metadata.year
+    val viewType = metadata.viewType
```

Comment:
There's getItemViewType() method and getViewTypeCount(), designed for this specifically and used for internal optimisations as well. Let's use it instead.